components/esp_http_server/src/httpd_ws.c

```diff
@@ -0,0 +1,311 @@
+    if (httpd_req_get_hdr_value_str(req, "Sec-WebSocket-Key", sec_key_encoded, sizeof(sec_key_encoded)) != ESP_OK) {
+        ESP_LOGW(TAG, LOG_FMT("Cannot find client key"));
+        return ESP_ERR_NOT_FOUND;
+    }
+
+    /* Prepare server key (Sec-WebSocket-Accept), concat the string */
+    char server_key_encoded[33] = { '\0' };
+    uint8_t server_key_hash[20] = { 0 };
+    char server_raw_text[sizeof(sec_key_encoded) + sizeof(ws_magic_uuid) + 1] = { '\0' };
+
+    strcpy(server_raw_text, sec_key_encoded);
+    strcat(server_raw_text, ws_magic_uuid);
+
+    ESP_LOGD(TAG, LOG_FMT("Server key before encoding: %s"), server_raw_text);
+
+    /* Generate SHA-1 first and then encode to Base64 */
+    size_t key_len = strlen(server_raw_text);
+    mbedtls_sha1_ret((uint8_t *)server_raw_text, key_len, server_key_hash);
+
+    size_t encoded_len = 0;
+    mbedtls_base64_encode((uint8_t *)server_key_encoded, sizeof(server_key_encoded), &encoded_len,
+                          server_key_hash, sizeof(server_key_hash));
+
+    ESP_LOGD(TAG, LOG_FMT("Generated server key: %s"), server_key_encoded);
+
+    /* Prepare the Switching Protocol response */
+    char tx_buf[192] = { '\0' };
+    int fmt_len = snprintf(tx_buf, sizeof(tx_buf),
+                           "HTTP/1.1 101 Switching Protocols\r\n"
+                           "Upgrade: websocket\r\n"
+                           "Connection: Upgrade\r\n"
+                           "Sec-WebSocket-Accept: %s\r\n\r\n", server_key_encoded);
+    if (fmt_len < 0 || fmt_len > sizeof(tx_buf)) {
+        ESP_LOGW(TAG, LOG_FMT("Failed to prepare Tx buffer"));
+        return ESP_FAIL;
+    }
+
+    /* Send off the response */
+    if (httpd_send(req, tx_buf, fmt_len) < 0) {
+        ESP_LOGW(TAG, LOG_FMT("Failed to send the response"));
+        return ESP_FAIL;
+    }
+
+    return ESP_OK;
+}
+
+static esp_err_t httpd_ws_check_req(httpd_req_t *req)
+{
+    /* Probe if input parameters are valid or not */
+    if (!req || !req->aux) {
+        ESP_LOGW(TAG, LOG_FMT("Argument is null"));
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    /* Detect handshake - reject if handshake was NOT YET performed */
+    struct httpd_req_aux *req_aux = req->aux;
+    if (!req_aux->sd->has_ws_handshake_done) {
+        ESP_LOGW(TAG, LOG_FMT("State is invalid - No handshake performed"));
+        return ESP_ERR_INVALID_STATE;
+    }
+
+    return ESP_OK;
+}
+
+static esp_err_t httpd_ws_mask_payload(uint8_t *payload, size_t len, const uint8_t *mask_key)
+{
+    if (len < 1 || !payload) {
+        ESP_LOGW(TAG, LOG_FMT("Invalid payload provided"));
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    for (size_t idx = 0; idx < len; idx++) {
+        payload[idx] = (payload[idx] ^ mask_key[idx % 4]);
+    }
+
+    return ESP_OK;
+}
+
+esp_err_t httpd_ws_recv_pkt(httpd_req_t *req, httpd_ws_pkt_t *pkt, size_t max_len, uint32_t timeout_sec)
+{
+    esp_err_t ret = httpd_ws_check_req(req);
+    struct httpd_req_aux *aux = req->aux;
```

Comment:
Return value of httpd_ws_check_req() should be checked before reading aux, as req may be NULL.

Reply:
@shahpiyushv Oh sorry, got a logic issue here